### PR TITLE
Add a function to create XYB jpegs and add it to benchmark_xl.

### PR DESCRIPTION
**Pull request by szabadka**
The quantization matrix and zero-bias logic is influenced by the adaptive quant field, based on the work of veluca93@.

Benchmark results:
```
Encoding            kPixels    Bytes          BPP  E MP/s  D MP/s     Max norm        pnorm       BPP*pnorm   Bugs
------------------------------------------------------------------------------------------------------------------
jpeg:q95              13270  7112433    4.2876819  17.799  45.441   1.72006321   0.48375988  2.074208483263      0
jpeg:q82              13270  3405650    2.0530730  21.529  54.414   3.46519852   0.93051614  1.910417529264      0
jpeg:libjxl:d1.0      13270  3444711    2.0766206   4.408  31.047   2.07481623   0.65874454  1.367962491749      0
jpeg:libjxl:d0.6      13270  4708255    2.8383395   4.028  20.722   1.50057375   0.48367719  1.372840076341      0
jxl:d1.0              13270  2681079    1.6162703   1.241   9.897   1.68643689   0.59894225  0.968052536559      0
```

### CODE REVIEW

**Comment by jonsneyers** [comment]
Could you compare this to non-XYB jpeg at similar bpp?

**Comment by szabadka** [comment]
> Could you compare this to non-XYB jpeg at similar bpp?

Updated the PR description. Added same-bpp and same-pnorm comparisons between libjpeg and xyb-jpeg

File: lib/extras/encode_jpeg.cc [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+#include "lib/jxl/opsin_params.h"
+HWY_BEFORE_NAMESPACE();
+namespace jxl {
+namespace extras {
+namespace HWY_NAMESPACE {
+
+void ComputeDCTCoefficients(const Image3F& opsin, const ImageF& qf,
+                            const FrameDimensions& frame_dim, const float* qm,
+                            std::vector<jpeg::JPEGComponent>* components) {
+  float qfmin, qfmax;
```
Comment (veluca93) [inline lib/extras/encode_jpeg.cc:34]: Why is this here?

File: lib/extras/encode_jpeg.cc [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+        TransformFromPixels(AcStrategy::Type::DCT,
+                            opsin.PlaneRow(c, 8 * by) + 8 * bx,
+                            opsin.PixelsPerRow(), dct, scratch_space);
+        jpeg::coeff_t* block = &coeffs[bix * kDCTBlockSize];
+        for (size_t iy = 0, i = 0; iy < 8; iy++) {
+          for (size_t ix = 0; ix < 8; ix++, i++) {
+            float coeff = 2040 * dct[i] * qmc[i];
+            float zero_bias = 0.5f * qfmax / qf.Row(by)[bx];
+            int cc = std::abs(coeff) < zero_bias ? 0 : std::round(coeff);
+            if (zero_bias > 1) {
```
Comment (veluca93) [inline lib/extras/encode_jpeg.cc:54]: I probably wrote this, so I don't have a lot of grounds to complain :P but some comments on what is going on here wouldn't hurt :)

File: lib/extras/encode_jpeg.cc [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+namespace {
+
+std::vector<uint8_t> CreateXybICCAppMarker() {
+  ColorEncoding c_xyb;
+  c_xyb.SetColorSpace(ColorSpace::kXYB);
+  c_xyb.rendering_intent = RenderingIntent::kPerceptual;
+  JXL_CHECK(c_xyb.CreateICC());
+  const auto& icc = c_xyb.ICC();
+  std::vector<uint8_t> icc_marker(17 + icc.size());
+  icc_marker[0] = 0xe2;
```
Comment (veluca93) [inline lib/extras/encode_jpeg.cc:87]: Can we add comments on what each byte/piece is doing?

File: lib/extras/encode_jpeg.cc [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+
+void AddJpegQuantMatrices(const ImageF& qf, float dc_quant,
+                          std::vector<jpeg::JPEGQuantTable>* quant_tables,
+                          float* qm) {
+  // Create a custom JPEG XL dequant matrix.
+  DequantMatrices dequant;
+  std::vector<QuantEncoding> encodings(DequantMatrices::kNum,
+                                       QuantEncoding::Library(0));
+  encodings[0] = QuantEncoding::DCT(
+      DctQuantWeightParams({{{{2000.0, -0.5, -0.5, -0.5, -0.5, -0.5}},
```
Comment (veluca93) [inline lib/extras/encode_jpeg.cc:105]: Where do the parameters come from?

File: lib/extras/encode_jpeg.cc [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+  out->marker_order.push_back(0xd9);
+
+  // Gather histograms.
+  jpeg::SerializationState ss;
+  JXL_CHECK(jpeg::ProcessJpeg(*out, &ss));
+
+  // Build DC Huffman codes and add them to DHT segment.
+  std::vector<Histogram> dc_histo;
+  std::vector<uint32_t> dc_tbl_indexes;
+  ClusterJpegHistograms(ss.dc_huff_table, 4, &dc_histo, &dc_tbl_indexes);
```
Comment (veluca93) [inline lib/extras/encode_jpeg.cc:285]: I have to admit it's entirely unclear to me what is going on here:
- who sets ss.dc_huff_table?
- why are we clustering to 4 tables and assigning clusters to scans sequentially? On the top of my head I don't see any reason why this should work.

Comment (veluca93) [reply lib/extras/encode_jpeg.cc:285]: Ah, nevermind, I didn't see the changes in jpeg_serialization_state, now it makes a lot more sense.

It's still probably not the optimal thing to do (in particular because clustering sort of assumes ANS, and ignores Huffman-specific things), but good enough for now.

File: lib/extras/encode_jpeg.cc (hunk 1 above)
Comment (szabadka) [reply lib/extras/encode_jpeg.cc:34]: qfmax is used in computing zero_bias

File: lib/extras/encode_jpeg.cc (hunk 2 above)
Comment (szabadka) [reply lib/extras/encode_jpeg.cc:58]: Done.

File: lib/extras/encode_jpeg.cc (hunk 4 above)
Comment (szabadka) [reply lib/extras/encode_jpeg.cc:114]: Added a comment.

File: lib/extras/encode_jpeg.cc (hunk 5 above)
Comment (szabadka) [reply lib/extras/encode_jpeg.cc:295]: Agreed, added a TODO to use JPEG-specific clustering.

**Review by veluca93 — APPROVED** [review]
